### ISSUE TITLE
morphology: dilation follows scipy (Minkowski) for odd kernels but OpenCV for even kernels and origin; skimage differs. Which conventions to keep?

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4713]
## Summary

`kornia.morphology` follows **two** conventions at once. With an **odd** kernel it is bit-for-bit `scipy.ndimage` (the Minkowski / textbook convention: `dilation` reflects the kernel, `erosion` does not). With an **even** kernel, or any non-default `origin`, it is bit-for-bit **OpenCV**, which puts the result one pixel away from both scipy and scikit-image. None of this is written in the docstrings.

This issue collects the executed comparison against scipy, scikit-image, OpenCV and torchmorph. It asks the `kornia.morphology` author to rule on what kornia should keep before the batch-7 Convention blocks document it (#4620).

**Not reopening:** the reflection question itself was asked in #2466 (2023). The ruling there was that flipping the structuring element in `dilation` is the correct, textbook behaviour. This issue does not challenge that ruling. It records it with evidence. #2466 also asked that the reflection *"should be mentioned at least in the documentation"*, and that has not happened yet.

## Two textbook conventions

| convention | dilation | erosion | opening / closing | libraries |
|---|---|---|---|---|
| **Minkowski** (Matheron / Serra; Gonzalez & Woods; Wikipedia) | `δ_B f(x) = max_{b∈B} f(x − b)`: **reflects** `B` | `ε_B f(x) = min_{b∈B} f(x + b)`: no reflection | `γ = δ_B ε_B`, `φ = ε_B δ_B` with the same `B` | **scipy.ndimage**, torchmorph, **kornia (odd kernels)** |
| **Neighbourhood** (Soille, *Morphological Image Analysis*) | max of `f` over the window `B` at `x`: **no reflection** (= Minkowski dilation by `B̌`) | same as Minkowski | reflect **internally** so they stay openings/closings | **scikit-image** (`skimage/morphology/gray.py`: `opening` calls `mirror_footprint` before `dilation`; `pad_footprint` for even sizes) |
| *(neither)* | no reflection | no reflection | composed naively; for an asymmetric kernel `MORPH_OPEN` is **not** an opening (`opening(block) ≠ block`) | **OpenCV** |

All three libraries agree on erosion. For odd kernels, Minkowski and neighbourhood give the same opening. Their closings differ by `B ↔ B̌`. For the **even-kernel centre** no textbook fixes a choice. scipy and skimage each pick one, and those two agree with each other on the dilation probe below. kornia picks OpenCV's.

## Executed comparison

kornia `0.9.0rc1` (main `088586663`), torch 2.14.0, cpu float32; scipy 1.17.1, scikit-image 0.26.0, opencv-python-headless 5.0.0. Fixtures: a hot pixel in `1×7` at col 3, and a hot pixel in `6×7` at `(2, 3)`; the whole-image rows use `torch.rand(1, 1, 9, 11, generator=torch.Generator().manual_seed(0))`.

| probe | kornia | scipy | skimage | OpenCV | torchmorph¹ |
|---|---|---|---|---|---|
| `dilation`, kernel `[[0,1,1]]` | cols `{3,4}` | `{3,4}` | `{2,3}` | `{2,3}` | `{3,4}` |
| `1 − erosion(1 − x)`, `[[0,1,1]]` | `{2,3}` | `{2,3}` | `{2,3}` | `{2,3}` | `{2,3}` |
| `dilation`, `ones(2,2)` | rows `{2,3}` × cols `{3,4}` | `{1,2}×{2,3}` | `{1,2}×{2,3}` | `{2,3}×{3,4}` | `{1,2}×{2,3}` |
| `dilation`, `ones(3,3)`, `origin=[0,0]` / `anchor=(0,0)` | spreads to rows `0..2`, cols `1..3` (toward **negative** indices) | `origin=(-1,-1)` spreads to **positive** indices | — | same as kornia | scipy |
| 9×11 rand, odd kernels (`[[0,1,1]]`, L = `[[0,0,0],[0,1,1],[0,1,0]]`): dilation / erosion / opening / closing bit-equal to kornia | — | **all 4** | erosion, opening only | erosion only | — |
| 9×11 rand, `ones(2,2)`: dilation / erosion / opening / closing bit-equal to kornia | — | erosion only | none | **all 4** | — |
| `opening(3×4 block, [[0,1,1]]) == block` | ✅ | ✅ | ✅ | ❌ | — |
| non-flat SE `[0, .5, 1]` (additive), interior | `[…, 1.5, 2.0, …]` | same | n/a (flat only) | n/a | same (`structure`) |
| default border | `geodesic` (ignore outside) | `mode='reflect'` | `mode='reflect'`; `'ignore'` = kornia's `geodesic` | `BORDER_CONSTANT` with ±∞ | `mode='reflect'` |

¹ [intcomp/torchmorph](https://github.com/intcomp/torchmorph) at `861c66f` is a scipy.ndimage clone. It runs on CUDA only, so these values come from reading `grey_kernel.cu` plus a numpy emulation of its index arithmetic (the emulation matched scipy in 110/110 random cases). The CUDA kernel was not run.

### Where the mixed behaviour comes from

`dilation` reflects the kernel **values** (`neighborhood.flip((0, 1))`). The **window position** is set by `origin` through the padding `[origin[1], k_w − origin[1] − 1, origin[0], k_h − origin[0] − 1]`, which is OpenCV's anchor semantics and is not reflected with it. For an odd kernel with the default centre, reflecting the position changes nothing, so the result equals scipy. For an even kernel (default centre `k // 2`) or any off-centre `origin`, the unreflected position is exactly what OpenCV computes. That puts kornia one pixel from scipy, which reflects both the values and the position.

The design history agrees with this reading. #1030 (the bug that started the 2021 refactor) compared kornia against `scipy.ndimage.grey_erosion`, and #1084 is titled *"Make Morphology compatible with both OpenCV and Scipy"*. The even-kernel and `origin` results are the one place where "both" cannot hold at the same time.

## Reproduction (torch only)

```python
import torch
from kornia.morphology import dilation, erosion

x = torch.zeros(1, 1, 1, 7); x[..., 3] = 1
k = torch.tensor([[0.0, 1.0, 1.0]])
print(dilation(x, k)[0, 0, 0].tolist())            # [0, 0, 0, 1, 1, 0, 0]  (scipy: same; skimage / OpenCV: cols 2, 3)
print((1 - erosion(1 - x, k))[0, 0, 0].tolist())   # [0, 0, 1, 1, 0, 0, 0]  (all references agree)

y = torch.zeros(1, 1, 6, 7); y[..., 2, 3] = 1
print(dilation(y, torch.ones(2, 2))[0, 0].nonzero().tolist())
# [[2, 3], [2, 4], [3, 3], [3, 4]]   OpenCV: same; scipy and skimage: [[1, 2], [1, 3], [2, 2], [2, 3]]
print(dilation(y, torch.ones(3, 3), origin=[0, 0])[0, 0].nonzero().tolist())
# rows 0..2 x cols 1..3              OpenCV anchor=(0, 0): same
```

<details><summary>Reference generation snippet (scipy / skimage / OpenCV)</summary>

```python
import numpy as np, cv2
from scipy import ndimage as ndi
from skimage import morphology as sm

x = np.zeros((1, 7), np.float32); x[0, 3] = 1
k = np.array([[0, 1, 1]], np.uint8)
on = lambda a: np.argwhere(a > 0.5).tolist()
print(on(ndi.grey_dilation(x, footprint=k.astype(bool))))  # [[0, 3], [0, 4]]
print(on(sm.dilation(x, k)))                               # [[0, 2], [0, 3]]
print(on(cv2.dilate(x, k)))                                # [[0, 2], [0, 3]]

y = np.zeros((6, 7), np.float32); y[2, 3] = 1
e = np.ones((2, 2), np.uint8)
print(on(ndi.grey_dilation(y, footprint=e.astype(bool))))  # [[1, 2], [1, 3], [2, 2], [2, 3]]
print(on(sm.dilation(y, e)))                               # [[1, 2], [1, 3], [2, 2], [2, 3]]
print(on(cv2.dilate(y, e)))                                # [[2, 3], [2, 4], [3, 3], [3, 4]]
```
</details>

## Questions for the `kornia.morphology` author

1. **Reflection:** keep the Minkowski reflection in `dilation` (the #2466 ruling; = scipy, torchmorph, textbook)? The alternative is scikit-image's convention (no reflection in `dilation`, reflection inside `opening` / `closing`). That would change `dilation`, `closing`, `gradient` and `bottom_hat` for every asymmetric kernel.
2. **Even kernels and `origin`:** was OpenCV's anchor behaviour for even kernels and `origin` intended (the "compatible with OpenCV" half of #1084), or should they follow scipy like the odd case does? If they should follow scipy, the whole module becomes one convention, and kornia agrees with scipy and torchmorph on every row above except the border default.
3. **`origin` meaning:** today it is an index into the kernel (OpenCV's `anchor`, `[row, col]`), not scipy's offset from the centre. Keep it as an index and document it, whichever of 2 is chosen?
4. **Border default:** keep `geodesic` (= skimage `mode="ignore"`, = OpenCV's default) rather than scipy/skimage's `reflect`? Note that `border_type="reflect"` here is `torch.nn.functional.pad`'s reflect, which scipy and skimage call `"mirror"`.

## Classification

Behavior convention, undocumented. Intentional-variant candidate for the reflection (#2466 ruling), wart candidate for the even-kernel / `origin` placement pending the author's answer to question 2. Idiom: none of 1–9 as far as I can tell. The closest description is "the values follow one reference, the window position another", and I have not assigned it a number.

## Disposition

Reflection: `KEEP SEPARATE — intentional variant` (document in the batch-7 `dilation` Convention block, closing the documentation half of #2466). Even-kernel centre and `origin` placement: `MERGE IN WINDOW — behavior change` if the answer to question 2 is "follow scipy", otherwise `KEEP SEPARATE — intentional variant` and document. Either way nothing changes before the ruling, and any behavior change lands in the coordinated Repair & Deprecation window (#4620).

Other batch-7 findings (the finite `max_val = 1e4` stand-in for ±∞ leaking into outputs above `|x| = 1e4`, and integer input failing with an internal overflow error) are unrelated to the convention question and will be filed separately.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
@Manza12 sorry for bothering you, but could you please take a look as original author and tell us which conventions we should keep further, or maybe keep as is?

**Comment by Manza12** [comment]
At a first glance, the only problematic behavior is regarding the even kernels; in my opinion, the center should be (width//2, height//2).
Regarding the way passing the origin, I prefer to directly pass the index of the structuring element (which is natural) rather than the bizarre offset rule from SciPy.
I think reflection is OK.

I don't know if this is exactly what the issuer wants, since the purpose of the issue is drawn among AI generated text.

For the issuer: what is your proposal/question?

**Comment by ducha-aiki** [comment]
@Manza12 thank you! 

The question is - what should we do, and if I understand correctly, then the following is answered, please confirm:

  - reflection (of the dilation kernel) is OK
  - origin: keep as is
  - Even kernels: (width//2, height//2). 
  
Now we have the bug (I guess?) in dilation, it flips the kernel values but not the window position, while erosion already treats the index correctly. I will fix that. 


The question remaining:

- should we keep "geodesic" as border default? 

Context
  - geodesic (kornia's default): pixels outside the image don't exist. The max or min is taken only over the in-image pixels. The code does this by     padding with the ∓max_val stand-in for infinity. This is Soille's convention for a bounded image, skimage's mode="ignore", and OpenCV's default.
  - reflect (the scipy and skimage default): outside pixels are invented by mirroring the image, d c b a | a b c d, so the edge pixel is used twice.

**Comment by ducha-aiki** [comment]
Status of the batch-7 audit that this thread decided:

- Rulings applied: `dilation` keeps the Minkowski reflection, `origin` stays an index into the structuring element, the even-size default centre is `[k_h // 2, k_w // 2]`, and `geodesic` stays the default border.
- With those, the even-kernel / custom-origin offset is a bug in `dilation` (it reflected the kernel values but padded with the unreflected origin): #4733, fixed by #4737 (one line; even kernels now match scipy and scikit-image bit-for-bit, odd kernels at the default origin are unchanged).
- The Convention blocks, the Conventions-page section and the pins are #4743, stacked on #4737.
- Other findings from the same audit, filed separately: #4734 (`max_val=1e4` is a finite sentinel that leaks into the output and caps the convolution engine's accuracy), #4735 (bool / uint8 / int64 input), #4736 (validation messages).

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
